docs/function-inlining.md

```diff
@@ -0,0 +1,97 @@
+# No support for user inlining (yet)
+
+## Summary
+<!-- https://luau-lang.org/performance#function-inlining-and-loop-unrolling -->
+Luau's bytecode compiler is smart about inlining functions. We would rather focus on improving this automation than risk the chance of its misuse.
+
+## Motivation
+
+**context**: function inlining replaces function calls with the function's implementation to reduce overhead costs.
+
+Whether the goal is to improve instruction locality or perform context specific optimizations, <!-- force the compiler to optimize the function body in specific contexts --> function inlining can be a great resource for developers. At `-02` or higher, the compiler is able to automatically decide if functions ought to be inlined. In the currrent state of the language, exposing this to users directly could do more harm than good. Not only do these use cases only cater to a tiny subset of the language's users, it exposes our bytecode compiler to stack overflow errors when the larger subset who may not have a working understanding of the compiler adopts it. For Luau, this would happen at around the 5000th recursive call.
+
+```luau
+local function sum(lst: {number})
+  return sumAcc(0, 1, lst)
+end
+
+@inline
+local function sumAcc(acc: number, i: number, lst: {number})
+  return if i > #lst then acc else sumAcc(acc + lst[i], i + 1, lst)
+end
+
+local s = sum({1, 2, 3})
+```
+
+Consider this "inlined" recursive call, it attempts to sum up numbers in a list but it never terminates as `sum` is recursively called on `lst` which does not shrink. If Luau attempted to inline such a call, it would unroll this up until the 5000th recursive call and end up thrashing our stack. The stack will essentially grow beyond the system's stack limit and could potentially eat up resources meant for other processes.
```

Comment:
So, this function does terminate. If `i` is greater than the length of `lst`, return the value we've added up thus far. `i` is incremented on each call. As long as `sumAcc` was able to finish in fewer call frames than the available call stack, the function will successfully return the value.

Also, not really too sure about how inlining leads to unrolling, but unrolling shouldn't cause a stack overflow. Unrolling refers to an optimization that takes a loop that's statically known to only occur exactly N times every time, and just copy paste the body N amount of times.

```suggestion
Consider this "inlined" recursive call. Inlining a recursive function would mean either the bytecode size is infinite, or the generated code is effectively the same as a loop a la tail call optimization.
```

Reply:
do we not want to explain what the code does in detail?

Reply:
Nah. I leave that as an exercise left to the reader. I want the non-code part of the RFC to focus on the _why_ the `inline` attribute can't work, not necessarily _what_ the code does.